File: modules/core/src/main/java/kr/entree/enderwand/command/StandardCommand.java
package kr.entree.enderwand.command;

import kr.entree.enderwand.command.argument.Argument;
import kr.entree.enderwand.command.completer.TabCompleter;
import kr.entree.enderwand.command.executor.CommandExecutor;
import kr.entree.enderwand.command.sender.Sender;

import java.util.List;

/**
 * Created by JunHyung Lim on 2020-02-04
 */
public class StandardCommand<S extends Sender> implements Command<S>, CommandDetail {
    private final CommandExecutor<S> executor;
    private final TabCompleter<S> tabCompleter;
    private final String description;
    private final String permission;
    private final List<Argument<?>> arguments;

    public StandardCommand(
            CommandExecutor<S> executor,
            TabCompleter<S> tabCompleter,
            String description,
            String permission,
            List<Argument<?>> arguments
    ) {
        this.executor = executor;
        this.tabCompleter = tabCompleter;
        this.description = description;
        this.permission = permission;
        this.arguments = arguments;
    }

    @Override
    public List<String> tabComplete(CommandContext<S> ctx) {
        return tabCompleter.tabComplete(ctx);
    }

    @Override
    public void execute(CommandContext<S> ctx) {
        executor.execute(ctx);
    }

    @Override
    public String description() {
        return description;
    }

    @Override
    public String permission() {
        return permission;
    }

    @Override
    public List<Argument<?>> arguments() {
        return arguments;
    }
}

File: modules/core/kotlin/src/main/kotlin/kr/entree/enderwand/command/StandardCommand.kt
package kr.entree.enderwand.command

import kr.entree.enderwand.collection.Reader
import kr.entree.enderwand.command.argument.Argument
import kr.entree.enderwand.command.sender.Sender

/**
 * Created by JunHyung Lim on 2020-01-09
 */
fun Sender.checkPermission(permission: String) {
    if (!hasPermission(permission)) {
        throw NoPermissionException(permission)
    }
}

class StandardCommand<S : Sender>(
    private val executor: CommandContext<S, Reader<String>>.() -> Unit,
    private val tabCompleter: CommandContext<S, Reader<String>>.() -> List<String>,
    override var description: String = "",
    override var permission: String = "",
    override var arguments: List<Argument<*>> = emptyList()
) : Command<S>, CommandDetailed {
    override fun execute(ctx: CommandContext<S, Reader<String>>) {
        try {
            ctx.sender.checkPermission(permission)
            executor(CommandContext(ctx.sender, ctx.args))
        } catch (ex: ExecutorException) {
            ex.stack += "" to this
            throw ex
        } catch (ex: Exception) {
            throw executorExceptionOf(this, ex)
        }
    }

    override fun tabComplete(
        ctx: CommandContext<S, Reader<String>>
    ) = tabCompleter(ctx)
}